builtin/logical/transit/lock_manager.go

```diff
@@ -0,0 +1,337 @@
+package transit
+
+import (
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/hashicorp/vault/logical"
+)
+
+const (
+	shared    = false
+	exclusive = true
+)
+
+type lockManager struct {
+	// A lock for each named key
+	locks map[string]*sync.RWMutex
+
+	// A mutex for the map itself
+	lockMutex sync.RWMutex
+
+	// If caching is enabled, the map of name to in-memory policy cache
+	cache map[string]*Policy
+
+	// Used for global locking, and as the cache map mutex
+	globalMutex sync.RWMutex
+}
+
+func newLockManager(cacheDisabled bool) *lockManager {
+	lm := &lockManager{
+		locks: map[string]*sync.RWMutex{},
+	}
+	if !cacheDisabled {
+		lm.cache = map[string]*Policy{}
+	}
+	return lm
+}
+
+func (lm *lockManager) CacheActive() bool {
+	return lm.cache != nil
+}
+
+func (lm *lockManager) LockAll(name string) {
+	lm.globalMutex.Lock()
+	lm.LockPolicy(name, exclusive)
+}
+
+func (lm *lockManager) UnlockAll(name string) {
+	lm.UnlockPolicy(name, exclusive)
+	lm.globalMutex.Unlock()
+}
+
+func (lm *lockManager) LockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	if lock != nil {
+		// We want to give this up before locking the lock, but it's safe --
+		// the only time we ever write to a value in this map is the first time
+		// we access the value, so it won't be changing out from under us
+		lm.lockMutex.RUnlock()
+		if writeLock {
+			lock.Lock()
+		} else {
+			lock.RLock()
+		}
+		return
+	}
+
+	lm.lockMutex.RUnlock()
+	lm.lockMutex.Lock()
+
+	// Don't defer the unlock call because if we get a valid lock below we want
+	// to release the lock mutex right away to avoid the possibility of
+	// deadlock by trying to grab the second lock
+
+	// Check to make sure it hasn't been created since
+	lock = lm.locks[name]
+	if lock != nil {
+		lm.lockMutex.Unlock()
+		if writeLock {
+			lock.Lock()
+		} else {
+			lock.RLock()
+		}
+		return
+	}
+
+	lock = &sync.RWMutex{}
+	lm.locks[name] = lock
+	lm.lockMutex.Unlock()
+	if writeLock {
+		lock.Lock()
+	} else {
+		lock.RLock()
+	}
+}
+
+func (lm *lockManager) UnlockPolicy(name string, writeLock bool) {
+	lm.lockMutex.RLock()
+	lock := lm.locks[name]
+	lm.lockMutex.RUnlock()
+
```

Comment:
I guess, although if that ever happens it's horrendously broken and should be caught by every unit test.  :-)